rust_snuba/src/strategies/commit_log.rs

```diff
@@ -81,6 +92,105 @@ impl TryFrom<Commit> for KafkaPayload {
     }
 }
 
+struct ProduceMessage {
+    producer: Arc<dyn Producer<KafkaPayload>>,
+    topic: Arc<TopicOrPartition>,
+    skip_produce: bool,
+}
+
+impl ProduceMessage {
+    #[allow(dead_code)]
+    pub fn new(
+        producer: impl Producer<KafkaPayload> + 'static,
+        topic: TopicOrPartition,
+        skip_produce: bool,
+    ) -> Self {
+        ProduceMessage {
+            producer: Arc::new(producer),
+            topic: Arc::new(topic),
+            skip_produce,
+        }
+    }
+}
+
+impl TaskRunner<KafkaPayload, KafkaPayload> for ProduceMessage {
+    fn get_task(&self, message: Message<KafkaPayload>) -> RunTaskFunc<KafkaPayload> {
+        let producer = self.producer.clone();
+        let topic = self.topic.clone();
+        let skip_produce = self.skip_produce;
+
+        Box::pin(async move {
+            if skip_produce {
+                return Ok(message);
+            }
+
+            let mut retries = 3;
+            while retries > 0 {
+                match producer.produce(&topic, message.payload()) {
+                    Ok(_) => {
+                        return Ok(message);
+                    }
+                    Err(e) => {
+                        retries -= 1;
+                        sleep(Duration::from_millis(100));
+                        log::warn!("{}", e);
+                    }
+                }
+            }
+            panic!("Failed to produce message");
+        })
+    }
+}
+
+pub struct ProduceCommitLog {
+    inner: Box<dyn ProcessingStrategy<KafkaPayload>>,
+}
+
```

Comment:
What do we need this outer structure for ?

Reply:
All the strategies do this since the nested processing step doesn't have a statically known size. I think this is actually your code originally.

Reply:
>  I think this is actually your code originally.

I would not trust it to be doing the right thing.

Reply:
Still, it's not a statically sized thing.. do you have another suggestion?

Reply:
> All the strategies do this since the nested processing step doesn't have a statically known size.

Not sure why this helps. Let's say the strategy was not there as a wrapper, where would we encounter problems with processing steps not having a statically known size ?

Reply:
Actually, the inner thing is always RunTaskInThreads in this case. So I can change the annotation and get rid of the box